### PR TITLE
enhance: Handle rust error in c++

### PR DESCRIPTION
**Pull request by sunby**
https://github.com/milvus-io/milvus/issues/37930

### CODE REVIEW

**Comment by sunby** [comment]
[/run-cpu-e2e](https://github.com/milvus-io/milvus/issues/37930)

**Comment by sunby** [comment]
https://github.com/milvus-io/milvus/issues/37930

File: internal/core/thirdparty/tantivy/rust-array.h [hunk 1]
```diff
@@ -62,4 +68,38 @@ struct RustArrayWrapper {
...
     }
 };
+struct RustResultWrapper {
+    NO_COPY_OR_ASSIGN(RustResultWrapper);
+
+    RustResultWrapper() = default;
+    explicit RustResultWrapper(RustResult result) : result_(result) {
+    }
+
+    RustResultWrapper(RustResultWrapper&& other) noexcept {
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/rust-array.h:78]: It seems that RustResultWrapper is not a safe type.
The move construction operation on it make a double free.
I prefer to wrap those `ffi` type such as RustResult with unique_ptr to make it safe.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs [hunk 2]
```diff
@@ -20,10 +30,132 @@ impl RustArray {
...
+
 #[no_mangle]
 pub extern "C" fn free_rust_array(array: RustArray) {
     let RustArray { array, len, cap } = array;
     unsafe {
         Vec::from_raw_parts(array, len, cap);
     }
 }
+
+#[repr(C)]
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:57]: does the `ffi` type satisfy the rfc https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md?
Please add some unittest to make sure that it's ok at current compiler.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs [hunk 3]
```diff
@@ -20,10 +30,132 @@ impl RustArray {
...
+                error: create_string(&e.to_string()),
+            },
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn free_rust_result(result: RustResult) {
+    match result.value {
+        Value::RustArray(array) => {
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:127]: In my view, the pointer type should also be freed.

Comment (sunby) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:127]: > In my view, the pointer type should also be freed.

Ptr type is catched by c++ class and will be freed when dtor is called for now. So we don't free ptr here.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs (hunk 2 above)
Comment (sunby) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:57]: > does the `ffi` type satisfy the rfc https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md? Please add some unittest to make sure that it's ok at current compiler.

as cbindgen doc says `cbindgen also supports using repr(C)/repr(u8) on non-C-like enums (enums with fields). This gives a C-compatible tagged union layout, as [defined by this RFC 2195](https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md). repr(C) will give a simpler layout that is perhaps more intuitive, while repr(u8) will produce a more compact layout.`

this enum should be supported. But I will also add a ut to verify it.

**Comment by sunby** [comment]
rerun go-sdk

**Review by congqixia — APPROVED** [review]
/approve

**Pull request merged**